### PR TITLE
[Tabular] Convert ColumnTransformer in tabular NN from sklearn to onnx

### PR DESCRIPTION
**Pull request by liangfu**
*Issue #, if available:*

*Description of changes:*

This PR converts the ColumnsTransformer in tabular NN model from sklearn to onnx, so that the processor in TabularNeuralNetTorchModel can be accelerated via onnxruntime.

While the impact to accuracy is minimal,
```
NeuralNetTorch:
{'accuracy': 0.8365236974101751, 'balanced_accuracy': 0.7532995553694549, 'mcc': 0.5305019485882506, 'roc_auc': 0.8815480292353528, 'f1': 0.6332950631458094, 'precision': 0.6769759450171822, 'recall': 0.594909404659189}
NeuralNetTorch_ONNX:
{'accuracy': 0.8362166035418159, 'balanced_accuracy': 0.7523552495805498, 'mcc': 0.5291951976581161, 'roc_auc': 0.8813204856717612, 'f1': 0.6320147194112237, 'precision': 0.6768472906403941, 'recall': 0.5927523727351165}
```
, the impact to overall online inference is significant.

The relative speedups comparing to sklearn baseline are shown in the following figure.

image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by liangfu** [comment]
Here is the script to demonstrate the acceleration.
```python
import os, shutil
import numpy as np
import pandas as pd
import autogluon.core as ag
from autogluon.tabular import TabularDataset, TabularPredictor
from autogluon.core.utils.infer_utils import get_model_true_infer_speed_per_row_batch

def main():

    # Training time:
    train_data = TabularDataset('https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv')  # can be local CSV file as well, returns Pandas DataFrame
    test_data = TabularDataset('https://autogluon.s3.amazonaws.com/datasets/Inc/test.csv')  # another Pandas DataFrame
    label = 'class'  # specifies which column do we want to predict
    save_path = 'ag_hpo_models/'  # where to save trained models
    compiled_path = 'ag_hpo_models_compiled/'

    subsample_size = 1000  # subsample subset of data for faster demo, try setting this to much larger values
    if subsample_size is not None and subsample_size < len(train_data):
        train_data = train_data.sample(n=subsample_size, random_state=0)

    hyperparameters = {
        'NN_TORCH': {},
    }

    predictor = TabularPredictor(label=label, path=save_path).fit(
        train_data, hyperparameters=hyperparameters,
        fit_weighted_ensemble=False,
    )
    predictor.save()

    repeats = 10
    batch_sizes = [
        1,
        10,
        100,
        1000,
        10000,
        100000,
    ]

    infer_dfs = dict()
    infer_transform_dfs = dict()
    silent=False
    test_data = test_data.head(10)

    for batch_size in batch_sizes:
        predictor = TabularPredictor.load(path=save_path)
        if os.path.exists(compiled_path):
            shutil.rmtree(compiled_path)
        predictor.clone(path=compiled_path)

        predictor = TabularPredictor.load(path=compiled_path)

        infer_df, time_per_row_transform = get_model_true_infer_speed_per_row_batch(
            data=test_data,
            predictor=predictor,
            batch_size=batch_size,
            repeats=repeats,
            persist_models=True,
            silent=silent)

        predictor = TabularPredictor.load(path=compiled_path)
        predictor.compile_models(models='all', compiler_configs={
            "NN_TORCH": {"compiler": 'onnx', 'batch_size': min(4096, batch_size)},
            "RF": {"compiler": 'onnx'},
        })

        infer_compiled_df, _ = get_model_true_infer_speed_per_row_batch(
            data=test_data,
            predictor=predictor,
            batch_size=batch_size,
            repeats=repeats,
            persist_models=True,
            silent=silent)

        infer_df = infer_df.reset_index()
        infer_compiled_df = infer_compiled_df.reset_index()
        for idx, name in enumerate(infer_compiled_df['model']):
            infer_compiled_df['model'][idx] = name + "_ONNX"
        infer_df = pd.concat([infer_df, infer_compiled_df])
        infer_dfs[batch_size] = infer_df
        infer_transform_dfs[batch_size] = time_per_row_transform

    for key in infer_dfs.keys():
        infer_dfs[key] = infer_dfs[key].reset_index()
        infer_dfs[key]['batch_size'] = key

    infer_df_full_transform = pd.Series(infer_transform_dfs, name='pred_time_test').to_frame().rename_axis('batch_size')
    infer_df_full_transform['pred_time_test_marginal'] = infer_df_full_transform['pred_time_test']
    infer_df_full_transform['pred_time_test_with_transform'] = infer_df_full_transform['pred_time_test']
    infer_df_full_transform = infer_df_full_transform.reset_index()

    infer_df_full = pd.concat([infer_dfs[key] for key in infer_dfs.keys()])

    infer_df_full_transform_include = infer_df_full_transform.copy()
    infer_df_full_transform_include['model'] = 'transform_features'
    infer_df_full = pd.concat([infer_df_full, infer_df_full_transform_include])

    infer_df_full = infer_df_full.sort_values(by=['batch_size'])
    infer_df_full = infer_df_full.reset_index(drop=True)

    print(infer_df_full)

    infer_df_full['rows_per_second'] = 1 / infer_df_full['pred_time_test']

    import matplotlib.pyplot as plt

    fig, ax = plt.subplots()
    fig.set_size_inches(12, 12)
    fig.set_dpi(300)

    plt.xscale('log')
    plt.yscale('log')

    models = list(infer_df_full['model'].unique())
    batch_sizes = list(infer_df_full['batch_size'].unique())
    for model in models:
        infer_df_model = infer_df_full[infer_df_full['model'] == model]
        ax.plot(infer_df_model['batch_size'].values, infer_df_model['rows_per_second'].values, label=model)

    ax.set(xlabel='batch_size', ylabel='rows_per_second',
           title='Rows per second inference throughput by data batch_size (AdultIncome)')
    ax.grid()
    ax.legend()
    fig.savefig('infer_speed.png', dpi=300)
    plt.show()

if __name__=='__main__':
    main()
```

**Comment by Innixma** [comment]
Why `test_data = test_data.head(10)` in script?

**Comment by liangfu** [comment]
> Why `test_data = test_data.head(10)` in script?

Nice catch, we can safely remove this line now. It was used for testing subset of the dataset that doesn't contain NaN.

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py [hunk 1]
```diff
@@ -590,16 +592,11 @@ def load(cls, path: str, reset_paths=True, verbose=True):
...
-            from .torch_network_modules import EmbedNet
-
-            # recreate network from architecture description
-            model.model = EmbedNet(problem_type=model.problem_type,
-                                   num_net_outputs=model._get_num_net_outputs(),
-                                   quantile_levels=model.quantile_levels,
-                                   architecture_desc=model._architecture_desc,
-                                   device=model.device)
             model._architecture_desc = None
             model.model = torch.load(model.path + model.params_file_name)
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py:596]: please use `os.path.join(model.path, model.params_file_name)` instead.

File: tabular/src/autogluon/tabular/models/tabular_nn/compilers/onnx.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        with open(path + "model.onnx", "wb") as f:
+            f.write(model.SerializeToString())
+        return path + "model.onnx"
+
+    @staticmethod
+    def load(path: str) -> TabularNeuralNetTorchOnnxTransformer:
+        """Load from the path that contains an onnx file."""
+        import onnx
+
+        onnx_bytes = onnx.load(path + "model.onnx")
```
[the remark this answers is not in the document]
Comment (liangfu) [reply tabular/src/autogluon/tabular/models/tabular_nn/compilers/onnx.py]: Using `os.path.join` to be compatible with windows.

File: tabular/src/autogluon/tabular/models/tabular_nn/compilers/onnx.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+class InferenceSessionWrapper:
+    """
+    Wrap around InferenceSession in onnxruntime, since it cannot be pickled.
+    See https://github.com/microsoft/onnxruntime/issues/10097
+    """
+
+    def __init__(self, onnx_bytes):
+        import onnxruntime as ort
+
+        ort.set_default_logger_severity(4)
```
[the remark this answers is not in the document]
Comment (liangfu) [reply tabular/src/autogluon/tabular/models/tabular_nn/compilers/onnx.py]: Removed this hard-coded line here.

File: tabular/src/autogluon/tabular/models/tabular_nn/compilers/native.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    @staticmethod
+    def save(model, path: str):
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        with open(path + 'model_native.pkl', 'wb') as fp:
+            fp.write(pickle.dumps(model))
+
+    @staticmethod
+    def load(path: str):
+        pkl = None
+        with open(path + 'model_native.pkl', 'rb') as fp:
```
[the remark this answers is not in the document]
Comment (liangfu) [reply tabular/src/autogluon/tabular/models/tabular_nn/compilers/native.py]: Using `os.path.join` to be compatible with windows.

File: tabular/src/autogluon/tabular/models/tabular_nn/compilers/native.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            A list of tuples containing shape and element type info, e.g. [((1, 14), np.float32),].
+            The list would be used as the input data for the model.
+            The compiler would optimize the model to perform best with the given input type.
+        """
+        AbstractNativeCompiler.save(model, path)
+
+    @staticmethod
+    def save(model, path: str):
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        with open(path + 'model_native.pkl', 'wb') as fp:
```
[the remark this answers is not in the document]
Comment (liangfu) [reply tabular/src/autogluon/tabular/models/tabular_nn/compilers/native.py]: Using `os.path.join` to be compatible with windows.

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py (hunk 1 above)
Comment (liangfu) [reply tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py]: Using `os.path.join` to be compatible with windows.

**Comment by liangfu** [comment]
I think this depends on https://github.com/onnx/sklearn-onnx/pull/938 to be part of a future release in skl2onnx package. Current latest release 1.13 doesn't include this change.

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py [hunk 6]
```diff
@@ -590,16 +592,11 @@ def load(cls, path: str, reset_paths=True, verbose=True):
...
-            from .torch_network_modules import EmbedNet
-
-            # recreate network from architecture description
-            model.model = EmbedNet(problem_type=model.problem_type,
-                                   num_net_outputs=model._get_num_net_outputs(),
-                                   quantile_levels=model.quantile_levels,
-                                   architecture_desc=model._architecture_desc,
-                                   device=model.device)
             model._architecture_desc = None
-            model.model = torch.load(model.path + model.params_file_name)
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py:602]: Naive question: Since we aren't doing anything with architecture_desc anymore here, do we still need to store it?

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py [hunk 7]
```diff
@@ -620,3 +617,40 @@ def _more_tags(self):
...
+        batch_size = compiler_configs.get("batch_size", self.max_batch_size)
+        compiler_fallback_to_native = compiler_configs.get('compiler_fallback_to_native', False)
+
+        self._compiler = self._get_compiler(compiler=compiler,
+                                            compiler_fallback_to_native=compiler_fallback_to_native)
+        if self._compiler is not None:
+            input_types = self._get_input_types(batch_size=batch_size)
+            self.processor = self._compiler.compile(model=(self.processor, self.model),
+                                                    path=self.path,
+                                                    input_types=input_types)
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py:656]: This looks like a lot of code dupe to sneak in the batch_size argument. Can we maybe make this more elegant to avoid code dupe?

File: tabular/tests/unittests/models/test_tabular_nn.py [hunk 8]
```diff
@@ -25,3 +26,35 @@ def test_tabular_nn_regression(fit_helper):
...
+    fit_args = dict(
+        hyperparameters={TabularNeuralNetTorchModel: {}},
+    )
+    dataset_name = 'covertype_small'
+    compiler_configs = {TabularNeuralNetTorchModel: {'compiler': 'onnx'}}
+    fit_helper.fit_and_validate_dataset(dataset_name=dataset_name, fit_args=fit_args,
+                                        compile_models=True, compiler_configs=compiler_configs)
+
+
+@pytest.mark.skip(reason="FunctionTransformer is not supported unless the transform function is None (= identity).")
```
Comment (Innixma) [inline tabular/tests/unittests/models/test_tabular_nn.py:51]: This doesn't need to skipped anymore right?

**Review by Innixma — APPROVED** [review]
LGTM, I can tell a lot of effort went into the specialized onnx compiler functions. Very nice speedup!

File: tabular/tests/unittests/models/test_tabular_nn.py (hunk 8 above)
Comment (liangfu) [reply tabular/tests/unittests/models/test_tabular_nn.py:51]: Actually, we still need to skip this, since we don't support converting `OneHotMergeRaresHandleUnknownEncoder` to onnx yet. It is required to handle columns with `onehot_features`. I think we can address this in a separate PR.

Comment (Innixma) [reply tabular/tests/unittests/models/test_tabular_nn.py:51]: so currently it will crash if someone calls compile with NN_TORCH on certain datasets? That is something we should be careful about.

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py (hunk 6 above)
Comment (liangfu) [reply tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py]: Nice catch!

I removed the architecture_desc, and everything works just fine.

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py (hunk 7 above)
Comment (liangfu) [reply tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py]: I put the call site of the compile() function into a `_compile` function, in order to reduce code duplication here.

File: tabular/tests/unittests/models/test_tabular_nn.py (hunk 8 above)
Comment (liangfu) [reply tabular/tests/unittests/models/test_tabular_nn.py]: That's a good point. I just added the support to the onehot encoder, so that 1) we don't need to skip this test case, 2) there shouldn't be a crash if someone calls compile with NN_TORCH on certain datasets.
